### DISCUSSION TITLE
Switch (focus) or open tab by URL command line

### DISCUSSION THREAD

**Discussion by gerardbosch** [tridactyl/tridactyl#4012] [Q&A] [answered]
Thanks for bringing Tridactyl :heart: 

Hi there! This is not strictly Tridactyl related question, but in some sense could fit Tridactyl –and I'm sure there's a lot of Firefox API knowledge in this room :)

**Problem:** How to focus an already open tab from command-line?

Firefox has built-in support to switch to an already open tab from any window when you type into address bar. Tridactyl also provides `taball` for the same purpose which is useful.

In the past I used [Shortkeys](https://www.shortkeys.app/) to define keyboard shortcuts to quickly jump to my most used sites (which I always have open), like Gmail, Google Calendar, Whatsapp, Keep,... For example, when I type:
- `gm` - Firefox jumps to the tab where `mail.google.com/*` is open
- `gw` - To jump to the tab where `web.whatsapp.com` is open
- ...

But this only worked, of course, while Firefox is focused. **Now I'd like to take this idea to the next level**, and be able to focus my mail (or whatever web app I use the most) from within any app, by using a hotkey software<sup id="a1">[1](#hotkey)</sup>.

I thought if I can run commands from a terminal, I would be able to achieve that, but after looking a little on the Firefox CLI, I only found options related to open new tabs, not to switch to an already open one or to list open tabs, e.g. `firefox --new-tab google.com`. Browsing the web I found what looks an old option to run remote commands, but I'm not sure this works these days: `firefox -remote "openURL(www.mozilla.org, new-tab)"`.

## Questions
1. Is there any native way to achieve what I'm describing? (run a command that opens an existent tab)
1. If not, could Tridactyl help somehow to workaround it? For example opening some URL in a new tab that Tridactyl then closes and makes the switch?
1. Finally, if achieving the above from outside Firefox was not possible at all, how can I configure Tridactyl to do the `gm`, `gw`,... (or whatever key sequence) focus the tab likewise Shortkeys? (Shortkeys allows focusing based on URL pattern, not tab title)
1. Any other idea?

Thanks beforehand!!

--
[<b id="hotkey">1</b>] At this time I'm using [KMonad](https://github.com/kmonad/kmonad), which allows you to fully program your keyboard, add layers, etc. Here you can find my WIP [configuration example](https://github.com/gerardbosch/dotfiles/blob/master/home/.config/kmonad/mi-keyboard.kbd) –but there's many other hotkey daemon programs for this. [↩](#a1)

**Comment by bovine3dom** [comment, marked as answer]
1. yeah, kinda, I think https://developer.mozilla.org/en-US/docs/Tools/Browser_Console#controlling_the_browser and https://developer.mozilla.org/en-US/docs/Tools/Remote_Debugging/Debugging_Firefox_Desktop , or maybe https://firefox-source-docs.mozilla.org/testing/marionette/Intro.html - but they give Firefox some scary red UI to let yo know that other processes can control it
2. yes, as you mentioned, https://github.com/tridactyl/tridactyl/issues/780 - but we'd need to rewrite the native messenger (again) to use the socket based / connection approach mentioned here: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Native_messaging#connectionless_messaging (I'm sure I talked about it a fair amount with someone recently but I can't work out where)
3. e.g. `:bind gm buffer google` where `google` is some fragment of the tab title, otherwise this PR https://github.com/tridactyl/tridactyl/pull/2927 might do what you want if you can remind me persistently enough to look at it that it gets merged some time this decade :)
4. I think 2) is the coolest way of doing it but it's unlikely to happen if it's left to me to do it ;)

Sorry for the delay in replying to this - I saw the email, thought "that's a cool question", and then promptly forgot about it

**Comment by bovine3dom** [comment]
https://github.com/tridactyl/tridactyl/issues/3914#issuecomment-925687938 ah, here is where we spoke about it recently
